gemini/src/main/java/com/techempower/gemini/path/AnnotationDispatcher.java

```diff
@@ -0,0 +1,550 @@
+    private final DispatchListener[]              listeners;
+
+    private ExecutorService  preinitializationTasks = Executors.newSingleThreadExecutor();
+    private Reflections      reflections            = null;
+
+    public AnnotationDispatcher(GeminiApplication application)
+    {
+        app = application;
+        handlers = new HashMap<>();
+        exceptionHandlers = new ExceptionHandler[]{};
+        prehandlers       = new Prehandler[]{};
+        listeners         = new DispatchListener[]{};
+
+        if (exceptionHandlers.length == 0)
+        {
+            throw new IllegalArgumentException("PathDispatcher must be configured with at least one ExceptionHandler.");
+        }
+
+        startReflectionsThread();
+    }
+
+    private void startReflectionsThread()
+    {
+        // Start constructing Reflections on a new thread since it takes a
+        // bit of time.
+        preinitializationTasks.submit(new Runnable() {
+            @Override
+            public void run() {
+                try
+                {
+                    reflections = PackageClassLoader.getReflectionClassLoader(app);
+                }
+                catch (Exception exc)
+                {
+                    // todo
+//                    log.log("Exception while instantiating Reflections component.", exc);
+                }
+            }
+        });
+    }
+
+    public void initialize() {
+        // Wait for pre-initialization tasks to complete.
+        try
+        {
+//            log.log("Completing preinitialization tasks.");
+            preinitializationTasks.shutdown();
+//            log.log("Awaiting termination of preinitialization tasks.");
+            preinitializationTasks.awaitTermination(5L, TimeUnit.MINUTES);
+//            log.log("Preinitialization tasks complete.");
+//            log.log("Reflections component: " + reflections);
+        }
+        catch (InterruptedException iexc)
+        {
+//            log.log("Preinitialization interrupted.", iexc);
+        }
+
+        // Throw an exception if Reflections is not ready.
+        if (reflections == null)
+        {
+            throw new ConfigurationError("Reflections not ready; application cannot start.");
+        }
+
+        register();
+    }
+
+    private void register() {
+//        log.log("Registering annotated entities, relations, and type adapters.");
+        try {
+            final ExecutorService service = Executors.newFixedThreadPool(1);
+
+            // @Path-annotated classes.
+            service.submit(new Runnable() {
+                @Override
+                public void run() {
+                    for (Class<?> clazz : reflections.getTypesAnnotatedWith(Path.class)) {
+                        final Path annotation = clazz.getAnnotation(Path.class);
+
+                        try {
+                            handlers.put(annotation.value(),
+                                    new AnnotationHandler(annotation.value(),

```

Comment:
I realize this complicates things, but you could support segment variables in the class-level `@Path`. I've seen this used for things like:

```java
@Path("/{lang}/foo")
public FooResource {
  @GET
  @Path("/bar/{id}")
  public Map getBarById(@PathParam("lang") String language, 
                        @PathParam("id") UUID id) {
    // ...
  }
}
```

Reply:
Yeah, definitely possible. This entire thing is a WIP, so if I get it working with the basics, I can expand on it like that. I opted for leaving static for this round because of just how much work went into getting variable capture working in a single handler (the tree searching is complicated by precedence rules between full URI matching, variable URI matching, and wildcard-at-any-depth matching).